src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java

```diff
@@ -369,35 +369,37 @@ private static void removeBuildFromNoBCList(Build build) {
 
     // trigger any builds whose watch events arrived before the
     // corresponding build config watch events
-    public synchronized static void flushBuildsWithNoBCList() {
-        boolean anyRemoveFailures = false;
-        for (Build build : buildsWithNoBCList.values()) {
-          WorkflowJob job = getJobFromBuild(build);
-          if (job != null) {
-            try {
-              logger.info("triggering job run for previously skipped build "
-                + build.getMetadata().getName());
-              triggerJob(job, build);
-            } catch (IOException e) {
-              logger.log(Level.WARNING, "flushBuildsWithNoBCList", e);
-            }
-            try {
-              removeBuildFromNoBCList(build);
-            } catch (Throwable t) {
-                // TODO
-                // concurrent mod exceptions are not suppose to occur
-                // with concurrent hash set; this try/catch with log 
-                // and the anyRemoveFailures post processing is a bit
-                // of safety paranoia until this proves to be true
-                // over extended usage ... probably can remove at some
-                // point
-                anyRemoveFailures = true;
-                logger.log(Level.WARNING, "flushBuildsWithNoBCList", t);
+    public static void flushBuildsWithNoBCList() {
+        synchronized(buildsWithNoBCList) {
+            boolean anyRemoveFailures = false;
+            for (Build build : buildsWithNoBCList.values()) {
+              WorkflowJob job = getJobFromBuild(build);
+              if (job != null) {
+                try {
+                  logger.info("triggering job run for previously skipped build "
+                    + build.getMetadata().getName());
+                  triggerJob(job, build);
+                } catch (IOException e) {
+                  logger.log(Level.WARNING, "flushBuildsWithNoBCList", e);
+                }
+                try {
+                  removeBuildFromNoBCList(build);
+                } catch (Throwable t) {
+                    // TODO
+                    // concurrent mod exceptions are not suppose to occur
+                    // with concurrent hash set; this try/catch with log 
+                    // and the anyRemoveFailures post processing is a bit
+                    // of safety paranoia until this proves to be true
+                    // over extended usage ... probably can remove at some
+                    // point
+                    anyRemoveFailures = true;
+                    logger.log(Level.WARNING, "flushBuildsWithNoBCList", t);
+                }
+              }
             }
-          }
-        }
-        if (anyRemoveFailures && buildsWithNoBCList.size() > 0) {
-            buildsWithNoBCList.clear();
+            if (anyRemoveFailures && buildsWithNoBCList.size() > 0) {
+                buildsWithNoBCList.clear();
```

Comment:
buildsWithNoBCList  is ConcurrentHashMap, thus:
`(More formally, an update operation for a given key bears a happens-before relation with any (non-null) retrieval for that key reporting the updated value.) For aggregate operations such as putAll and clear, concurrent retrievals may reflect insertion or removal of only some entries.`

Shouldn't we the "clear" synchronised ?

Reply:
ah good catch ... will update @akram

Reply:
oh wait ... it is synchronized ... look at line 373:  `synchronized(buildsWithNoBCList) {`

the clean call is in that synchronized block

so we are good here